[CI] Use LocalExecutor if kubeconfig is defined

nebril opened the pull request:
Executor interface is implemented by both LocalExecutor and
SSHMeta

This PR should not change the way tests are run on CI, but introduces code which will allow to switch to local kubectl execution as soon as we allow access to vagrant k8s cluster from host and are able to gather logs from all nodes via kubectl instead of ssh commands.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8683)
<!-- Reviewable:end -->

nebril committed:
[CI] Use LocalExecutor if kubeconfig is defined

Executor interface is implemented by both LocalExecutor and
SSHMeta

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

nebril committed:
[CI] Move vagrant start script to separate file

Add retrieval of kubeconfig to start script

Gather logs from pod instead of from host

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

eloycoto reviewed (commented):
This is super cool :-) Thanks!

nebril committed:
[CI] Move BasePath to SSHMeta and Kubectl structs

BasePath was a global variable in test suite. It was moved into
Executor interface, so that tests may access it base on the executor
they are using.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

nebril committed:
[CI] Run dig in pod instead of on host

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

nebril committed:
[CI] Move curl calls to utility pod

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

nebril committed:
[CI] Don't kill netperf process in chaos tests

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>

ianvernon commented on test/get-vagrant-kubeconfig.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
This have `-e` as a flag so that if it fails, provisioning fails.

ianvernon commented on ginkgo.Jenkinsfile:
```diff
@@ -116,6 +116,7 @@ pipeline {
                         TESTDIR="${GOPATH}/${PROJ_PATH}/test"
                         NETNEXT="true"
                         K8S_VERSION="1.11"
+						KUBECONFIG="vagrant-kubeconfig"
```
Indentation is off.

ianvernon commented on ginkgo.Jenkinsfile:
```diff
@@ -144,13 +149,21 @@ pipeline {
                         GOPATH="${WORKSPACE}/${TESTED_SUITE}-gopath"
                         TESTDIR="${GOPATH}/${PROJ_PATH}/test"
                         K8S_VERSION="1.15"
+						KUBECONFIG="vagrant-kubeconfig"
```
Here as well.

ianvernon commented on ginkgo.Jenkinsfile:
```diff
@@ -144,13 +149,21 @@ pipeline {
-                            sh 'cd ${TESTDIR}; vagrant destroy k8s1-1.15 k8s2-1.15 --force'
-                            sh 'cd ${TESTDIR}; vagrant up k8s1-1.15 k8s2-1.15 --provision'
+                            dir("${TESTDIR}") {
+                                sh 'cd ${TESTDIR}; vagrant destroy k8s1-1.15 k8s2-1.15 --force'
```
Is it possible to put this all into a script? Would be a bit easier to read than having to do a bunch of `sh` commands in a row in the Jenkinsfile.

ianvernon commented on test/k8sT/manifests/log-gatherer.yaml:
```diff
@@ -0,0 +1,59 @@
+        - "10000"
+        command:
+        - sleep
+        image: nebril/systemd:latest
```
Create a TODO to have this be made to be part of the Cilium image repo.

ianvernon commented on test/provision/k8s_install.sh:
```diff
@@ -268,12 +268,13 @@ sudo rm -rfv /var/lib/kubelet
 if [[ "${HOST}" == "k8s1" ]]; then
-    if [[ "${SKIP_K8S_PROVISION}" == "false" ]]; then
-      echo "${KUBEADM_CONFIG}" | envtpl > /tmp/config.yaml
+	if [[ "${SKIP_K8S_PROVISION}" == "false" ]]; then
```
Indentation looks off.

ianvernon commented on test/helpers/cons.go:
```diff
@@ -43,11 +43,14 @@ var (
 
 	// CiliumBasePath is the absolute path to the cilium source repository
```
This comment is wrong - it is a relative path now.

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -30,19 +30,19 @@ var _ = Describe("K8sPolicyTest", func() {
-		knpAllowIngress      = helpers.ManifestGet("knp-default-allow-ingress.yaml")
-		knpAllowEgress       = helpers.ManifestGet("knp-default-allow-egress.yaml")
-		cnpMatchExpression   = helpers.ManifestGet("cnp-matchexpressions.yaml")
+		demoPath             = ""
```
Why not just have these all by `demoPath, l3Policy ... string`?

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -840,7 +840,7 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
 	// If it fails we want to know if it's because of connection cannot be
 	// established or DNS does not exist.
-	digCMDFallback := "dig +tcp %s @%s"
+	//digCMDFallback := "dig +tcp %s @%s"
```
Why is this commented out?

nebril replied:
Edited - I didn't know the purpose of this command, it's back now.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -860,12 +860,30 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
 		if serviceIP == v1.ClusterIPNone {
-			res := kub.ExecShort(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
-			_ = kub.ExecShort(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
+			ress, err := kub.ExecInPods(context.TODO(), "kube-system", "k8s-app=cilium-test-logs", fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP), true)
```
Create a context with a reasonable timeout (e.g., 1 minute) so that the test will fail if the command gets stuck.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -860,12 +860,30 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
+				res = v
+				break
+			}
+			//_ = kub.ExecShort(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
```
Why is this removed?

nebril replied:
same as line 843

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -860,12 +860,30 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
 		}
 		log.Debugf("service is not headless; checking whether IP retrieved from DNS matches the IP for the service stored in Kubernetes")
-		res := kub.ExecShort(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+		ress, err := kub.ExecInPods(context.TODO(), "kube-system", "k8s-app=cilium-test-logs", fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP), true)
```
helpers.Kubesystem - perhaps hardcode the `k8s-app=cilium-test-logs` elsewhere as well - I want to avoid hardcoding of strings to make the code more maintainable.

ianvernon replied:
Use a context with a timeout here as well.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -860,12 +860,30 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
+			return false
+		}
+		var res *CmdRes
+		for _, v := range ress {
```
What is the rationale behind this loop? Is it because we only need one of the results at all?

nebril replied:
Changed it to different helper function which I created after this code.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -879,7 +897,7 @@ func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) er
 		if !strings.Contains(serviceIPFromDNS, serviceIP) {
 			logger.Debugf("service IP retrieved from DNS (%s) does not match the IP for the service stored in Kubernetes (%s)", serviceIPFromDNS, serviceIP)
-			_ = kub.ExecShort(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
+			//_ = kub.ExecShort(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
```
Why is this commented out?

nebril replied:
It's back now.

ianvernon commented on test/k8sT/Chaos.go:
```diff
@@ -232,12 +232,11 @@ var _ = Describe("K8sChaosTest", func() {
 				helpers.DefaultNamespace,
 				netperfClient,
-				fmt.Sprintf("netperf -l 300 -t TCP_STREAM -H %s", podsIps[netperfServer]))
+				fmt.Sprintf("netperf -l 60 -t TCP_STREAM -H %s", podsIps[netperfServer]))
```
How does reducing the timeout relate to the killing of the process?

nebril replied:
Killing the process via `cancel` causes the netperf run via `os/exec` to return error, which causes the test to fail. I am not sure how to stop netperf non-destructively at this point, so I reduced the timeout and simply wait for it to finish. I agree that it's not ideal.

ianvernon commented on test/helpers/cons.go:
```diff
@@ -41,13 +41,16 @@ var (
 	CiliumStartTimeout = 100 * time.Second
 
-	// CiliumBasePath is the absolute path to the cilium source repository
+	// CiliumBasePath is the path to the cilium source repository
```
Fix this discrepancy.

ianvernon commented on test/helpers/cons.go:
```diff
@@ -41,13 +41,16 @@ var (
-	CiliumBasePath = "/home/vagrant/go/src/github.com/cilium/cilium"
+	CiliumBasePathHost = "../"
+
+	CiliumBasePathVM = "../"
```
Fix this discrepancy.

ianvernon reviewed (changes requested):
Looks reasonable overall. I think this is a great first step. I have some misc. nits / feedback, but no big concerns.

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -105,39 +108,59 @@ func GetCurrentIntegration() string {
-	if node == nil {
-		ginkgoext.Fail(fmt.Sprintf("Cannot connect to vmName  '%s'", vmName), 1)
-		return nil
+	if config.CiliumTestConfig.Kubeconfig == "" {
```
Since `config.CiliumTestConfig.Kubeconfig` is the only global here, maybe just pass it in? When we come to change this (again) we'll have 1 less thing to think about.

nebril replied:
Not sure about that, we don't leak config to tests themselves right now, and this change would need for tests to be aware of user config. If you insist I can make this change.

raybejjani replied:
But by doing this we are leaking, no? The import tree now includes the config package. It's not that important, though.

raybejjani commented on test/k8sT/Chaos.go:
```diff
@@ -228,12 +232,11 @@ var _ = Describe("K8sChaosTest", func() {
 			restartCilium()
 
 			By("Stopping netperf client test")
-			cancel()
```
Unless the double cancel was causing an issue, it would be better to leave it out for another fixups PR.

nebril replied:
This cancel was killing the kubectl process as opposed to closing the ssh connection (previous version). 

As I wrote above: Killing the process via cancel causes the `kubectl exec netperf` run via os/exec to return error, which causes the test to fail. I am not sure how to stop netperf non-destructively at this point, so I reduced the timeout and simply wait for it to finish. I agree that it's not ideal.

raybejjani commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -34,8 +34,8 @@ var _ = Describe("K8sKafkaPolicyTest", func() {
-		l7Policy            = helpers.ManifestGet("kafka-sw-security-policy.yaml")
-		demoPath            = helpers.ManifestGet("kafka-sw-app.yaml")
+		l7Policy            = ""
+		demoPath            = ""
```
Maybe comment that these are set in BeforeAll below. It's more than a page away and may not be obvious.

raybejjani commented on test/k8sT/manifests/log-gatherer.yaml:
```diff
@@ -0,0 +1,59 @@
+        - "10000"
+        command:
+        - sleep
+        image: nebril/systemd:latest #TODO: move to cilium org
```
I imagine the image is pretty stable now, so we could move it before we merge if you want.

nebril replied:
Don't have access rights to our dockerhub (security!). Will change this in followup PR

raybejjani replied:
neither do I! We'll do it later.

raybejjani commented on test/test_suite_test.go:
```diff
@@ -264,9 +264,9 @@ var _ = BeforeAll(func() {
 			}
 		}
-		kubectl := helpers.CreateKubectl(helpers.K8s1VMName(), logger)
+		var kubectl *helpers.Kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
```
I'm pretty sure `kubectl :=` is the same as `var kubectl *helpers.Kubectl`. Was the change just to make something more obvious?

raybejjani commented on test/vagrant-ci-start.sh:
```diff
@@ -0,0 +1,12 @@
+
+vagrant ssh k8s1-${K8S_VERSION} -- sudo cat /etc/kubernetes/admin.conf
+
+kubectl get nodes
```
Perhaps having an echo explaining what this line is for will help when it breaks. The rest of the steps may also be improved with some logging.

raybejjani commented on test/vagrant-ci-start.sh:
```diff
@@ -0,0 +1,12 @@
+
+./get-vagrant-kubeconfig.sh > vagrant-kubeconfig
+
+vagrant ssh k8s1-${K8S_VERSION} -- sudo cat /etc/kubernetes/admin.conf
```
Is this just for debugging so it's in the logs somewhere? If so, wouldn't `cat vagrant-kubeconfig` be more correct?

nebril replied:
`cat vagrant-kubeconfig` will give already processed kubeconfig, this was added to show what is actually in source file. I can delete this, as it was for debugging.

raybejjani commented on test/helpers/cilium.go:
```diff
@@ -336,7 +336,11 @@ func (s *SSHMeta) GetEndpointsNames() ([]string, error) {
 // TODO: this can just be a constant; there's no need to have a function.
 func (s *SSHMeta) ManifestsPath() string {
-	return fmt.Sprintf("%s/runtime/manifests/", BasePath)
+	return fmt.Sprintf("%s/runtime/manifests/", s.basePath)
```
The commit message for `[CI] Move BaseString to SSHMeta and Kubectl structs` is wrong, since it's BasePath. Also, since it's a more general change, could you add some explanation in the commit ?

nebril replied:
done

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -110,6 +110,7 @@ func GetCurrentIntegration() string {
 type Kubectl struct {
 	Executor
 	*serviceCache
+	BasePath string
```
Since we have `.BasePath()` on the SSHMeta type, maybe it would be easier to add that to Executor and use that where we use this field?

nebril replied:
done

raybejjani commented on test/helpers/kubectl.go:
```diff
@@ -1053,8 +1086,9 @@ func (kub *Kubectl) generateCiliumYaml(options []string, filename string) error
 
 	// TODO GH-8753: Use helm rendering library instead of shelling out to
 	// helm template
+	helmTemplate := filepath.Join(kub.BasePath(), HelmTemplate)
```
@nebril I noticed that we were using .Join but we already have a `GetFilePath` that does it for us on the kubectl type.
